server/src/main/java/org/elasticsearch/rest/ChunkedZipResponse.java

```diff
@@ -0,0 +1,485 @@
+                    // longer an active AvailableChunksZipResponseBodyPart, so any concurrent calls to enqueueEntry() at this point will now
+                    // spawn a new AvailableChunksZipResponseBodyPart to take our place.
+                    isPartComplete = true;
+                    assert getNextPartListener == null;
+                    assert nextAvailableChunksListener != null;
+                    getNextPartListener = nextAvailableChunksListener;
+                } else {
+                    // The current entry is complete, and the first part of the next entry is already available, so we start sending its
+                    // chunks too. This means we're still the unique active AvailableChunksZipResponseBodyPart. We re-use this
+                    // AvailableChunksZipResponseBodyPart instance rather than creating a new one to avoid unnecessary allocations.
+                    final var nextEntry = entryQueue.poll();
+                    assert nextEntry != null;
+                    zipEntry = nextEntry.zipEntry();
+                    bodyPart = nextEntry.firstBodyPart();
+                    currentEntryReleasable = nextEntry.releasable();
+                }
+            } else {
+                // The current entry has more parts to come, but we have reached the end of the current part, so we assume that the next
+                // part is not yet available and therefore must pause transmission. This means we are no longer an active
+                // AvailableChunksZipResponseBodyPart, but also another call to enqueueEntry() won't create a new
+                // AvailableChunksZipResponseBodyPart because the current entry is still counted in queueLength:
+                assert queueLength.get() > 0;
+                // Instead, we create a new active AvailableChunksZipResponseBodyPart when the next part of the current entry becomes
+                // available. It doesn't affect correctness if the next part is already available, it's just a little less efficient to make
+                // a new AvailableChunksZipResponseBodyPart in that case. That's ok, entries can coalesce all the available parts together
+                // themselves if efficiency really matters.
+                isPartComplete = true;
+                assert getNextPartListener == null;
+                getNextPartListener = SubscribableListener.newForked(
+                    l -> bodyPart.getNextPart(l.map(p -> new AvailableChunksZipResponseBodyPart(null, p)))
+                );
+            }
+        }
+
+        private void finishResponse(ArrayList<Releasable> releasables) throws IOException {
+            assert zipEntry == null;
+            assert entryQueue.isEmpty() : entryQueue.size();
+            zipOutputStream.finish();
+            isPartComplete = true;
+            isLastPart = true;
+            transferCurrentEntryReleasable(releasables);
```

Comment:
Similarly, in practice I think it can't, it can only propagate exceptions thrown from the underlying stream, and in this case the underlying stream is entirely in-memory and won't throw anything either. Note that we assert that no exceptions are thrown while `targetStream != null`. But even if it did throw, that'd still be safe, we keep a hold of the resource in `currentEntryReleasable` and release it in `drainQueue` on the cleanup path.